File: src/main/java/core/WalletApplicationRunner.java
package src.main.java.core;

import static src.main.java.utils.ExceptionResolverUtils.tryCatchLogger;

import java.util.Scanner;

public class WalletApplicationRunner implements Runnable {
    private final Wallet wallet;
    private boolean isRunning;

    public static final String MENU_MESSAGE = """
            Wallet Menu:
            1. Open
            2. Close
            3. Get Balance
            4. Withdraw
            5. Store Card
            6. Get Card
            7. Get cards
            8. Deposit
            0. Exit
            Enter your choice:""";
    private final Scanner scanner;

    public WalletApplicationRunner(Wallet wallet, Scanner scanner) {
        assert (wallet != null) : " wallet must not be null for the runner to work well";
        assert (scanner != null) : "Scanner is invalid. User inputs are unreadable";
        this.wallet = wallet;
        this.scanner = scanner;
        isRunning = false;
    }

    private void showMenu() {
        System.out.println(wallet.getState());
        System.out.println(MENU_MESSAGE);
    }

    private void handleChoice(int choice) {
        switch (choice) {
            case 1 -> wallet.open();
            case 2 -> wallet.close();
            case 3 -> finallySkippedTryCatchLogger(wallet::getBalance);
            case 4 -> {
                System.out.print("Enter withdrawal amount: $");
                double withdrawalAmount = scanner.nextDouble();
                finallySkippedTryCatchLogger(() -> wallet.withdraw(withdrawalAmount));
            }
            case 5 -> {
                scanner.nextLine(); // Consume newline
                System.out.print("Enter card name to store: ");
                String cardName = scanner.nextLine();
                finallySkippedTryCatchLogger(() -> wallet.storeCard(cardName));
            }
            case 6 -> {
                scanner.nextLine(); // Consume newline
                System.out.print("Enter card name to retrieve: ");
                String retrieveCardName = scanner.nextLine();
                finallySkippedTryCatchLogger(() -> wallet.getCard(retrieveCardName));
            }
            case 7 -> finallySkippedTryCatchLogger(() -> System.out.println("cards = " + wallet.getCards()));
            case 8 -> {
                scanner.nextLine(); // Consume newline
                System.out.print("Enter amount to deposit: $");
                double depositAmount = scanner.nextDouble();
                finallySkippedTryCatchLogger(() -> wallet.deposit(depositAmount));
            }
            case 0 -> stop();
            default -> System.out.println("Invalid choice. Please enter a valid option.");
        }
    }

    @Override
    public void run() {
        do {
            isRunning = true;
            showMenu();

            int choice = scanner.nextInt();

            handleChoice(choice);
        } while (isRunning);
    }

    public void stop() {
        isRunning = false;
        System.out.println("Closing Wallet Application");
        closeResources();
    }

    public void closeResources() {
        scanner.close();
    }

    private void finallySkippedTryCatchLogger(Runnable tryCb) {
        tryCatchLogger(tryCb, () -> {
        });
    }
}

File: core/WalletApplicationRunner.js
import tryCatchLogger from "../utils.js";

export default class WalletApplicationRunner {
    wallet;
    userInputReader;
    isRunning;

    constructor(wallet, userInputReader) {
        this.wallet = wallet;
        this.userInputReader = userInputReader;
        this.isRunning = false;
    }


    run() {
        this.isRunning = true;

        this.showMenu();

        this.userInputReader.question("", (answer) => {
            this.continuedTryCatchLogger(() => this.handleChoice(answer))
        })
    }

    stop() {
        this.isRunning = false
    }

    showMenu() {
        this.showWalletState();
        console.log(`
            Wallet Menu:
                    1. Open
                    2. Close
                    3. Get Balance
                    4. Withdraw
                    5. Store Card
                    6. Get cards
                    7. Get Card
                    8. Deposit
                    0. Exit
                    Enter your choice:
        `)
    }

    showWalletState = () => {
        console.log(`Wallet : {isOpen : ${this.wallet.isOpen}}`)
    }

    handleChoice(choice) {
        let number = parseInt(choice);
        if (isNaN(number)) {
            console.log("unknown Argument, try again")
            return;
        }
        switch (number) {
            case 1: {
                this.wallet.open();
                break;
            }
            case 2 : {
                this.wallet.close();
                break;
            }

            case 3 : {
                let balance = undefined;
                //catcher for consistency
                tryCatchLogger(() => {
                    balance = this.wallet.getBalance();
                }, () => {
                })
                if (balance !== undefined)
                    console.log(`current balance is ${balance}`);
                break;
            }

            case 4 : {
                this.userInputReader.question("enter amount to withdraw ; ", answer => {
                    this.continuedTryCatchLogger(() => {
                        if (isNaN(parseFloat(answer))) {
                            console.log("unknown Argument, try again")
                            return;
                        }
                        this.wallet.withdraw(answer)
                    });
                })
                break;
            }

            case 5 : {
                this.userInputReader.question("enter card name to store ; ", answer => {
                    this.continuedTryCatchLogger(() => this.wallet.storeCard(answer));
                })
                break;
            }
            case 6 : {
                tryCatchLogger(this.wallet.getCards, () => {
                });
                break;
            }
            case 7 : {
                this.userInputReader.question("enter card name to retrieve ; ", answer => {
                    this.continuedTryCatchLogger(() => this.wallet.getCard(answer));
                })
                break;
            }
            case 8 : {
                this.userInputReader.question("enter amount to deposit ; ", answer => {
                    this.continuedTryCatchLogger(() => {
                        if (isNaN(parseFloat(answer))) {
                            console.log("unknown Argument, try again")
                            return;
                        }
                        this.wallet.deposit(answer)
                    });
                })
                break;
            }
            case 0: {
                this.stop()
                break;
            }
        }
    }

    continue() {
        if (this.isRunning) {
            this.run();
        } else {
            console.log("Wallet application closed")
            process.exit(0)
        }
    }

    continuedTryCatchLogger = (tryCb) => {
        tryCatchLogger(tryCb, () => this.continue())
    }
}
